# Homework 1

## Problem 1
 optimization of the function $(x_1 - x_2)^2 + (x_2 + x_3 -2)^2 + (x_4 -1)^2 + (x_5 -1)^2$

In [16]:
import numpy as np
def eq1(x):
    return (x[0] - x[1])**2 + (x[1] + x[2] -2)**2 + (x[3] -1)**2 + (x[4] -1)**2



All X must be within the bounds of -10 and 10, This can be done using scipy.optimize bounds function which requires each bound to be defined in a matrix form.
$\begin{bmatrix} -10\\ -10\\ -10\\ -10\\ -10\\ \end{bmatrix} \leq x \leq \begin{bmatrix} 10\\ 10\\ 10\\ 10\\ 10\\ \end{bmatrix}$

In [17]:

from scipy.optimize import Bounds
bounds = Bounds(-10, 10)


There are also 3 linear constraints as follows
$ x_1 + 3 x_2 = 0$
$ x_3 + x_4 - 2 x_5 = 0$
$ x_2 - x_5 = 0$


In [18]:
from scipy.optimize import LinearConstraint
linear_constraint = LinearConstraint([[1, 3, 0, 0, 0],[0, 0, 1, 1, 2],[0, 1, 0, 0, 1]], [0, 0, 0],[0, 0,0])
from scipy.optimize import minimize
x0 = np.array([5, 1, -2, -4, 1])
res = minimize(eq1, x0,method='trust-constr', constraints= linear_constraint, options={'verbose':1}, bounds=bounds )
print(res.x)


`xtol` termination condition is satisfied.
Number of iterations: 54, function evaluations: 420, CG iterations: 69, optimality: 3.37e-08, constraint violation: 5.55e-17, execution time: 0.07 s.
[-0.48837208  0.16279069  0.58139534 -0.25581396 -0.16279069]


C:\Users\trevo\PycharmProjects\pythonProject\venv\lib\site-packages\scipy\optimize\_hessian_update_strategy.py:182: UserWarning: delta_grad == 0.0. Check if the approximated function is linear. If the function is linear better results can be obtained by defining the Hessian as zero instead of using quasi-Newton approximations.
  warn('delta_grad == 0.0. Check if the approximated '


With Initial values of: [3, -1, -1, -1, -1]
The minimum values are: [-0.4883721   0.1627907   0.58139533 -0.25581393 -0.1627907 ]
With Initial values of: [5, 1, -2, -4, 1]
The minimum values are: [-0.48837208  0.16279069  0.58139534 -0.25581396 -0.16279069]
Taking a look at this, a different guess did not significantly change independent values for the minimized solution


## Problem 2
x and b $\in \mathbb{R}^n$ be vectors and A $\in \mathbb{R}^{n \times n}$ be a square matrix. Define
 $ f: \mathbb{R}^n \rightarrow \mathbb{R}$  as $ f(x) = b^Tx + x^T Ax$
a) What is the gradient and Hessian of f(x) with respect to x
$g(x) = b + 2Ax$
$H(x) = 2A$

b) Derive the first and second order Taylor's approximations of of $f(x)$ at $x=0$. Are the approximations exact?
First Order:
$f(x) = f(x_0) + \nabla f|^T_{x_0}(x-x_0)$
$f(x) = 0 + b^T(x)$
$f(x) =b^T(x)$
This is not an exact approximation. $f(x)$ is a second order polynomial so a first order approximation cannot be exact.

Second Order:
$f(x) = f(x_0) + \nabla f|^T_{x_0}(x-x_0) + \frac{1}{2}(x-x_0)^TH|_{x_0}(x-x_0)$
$f(x) = 0 +b^Tx + 2Ax$
$f(x) = b^Tx + 2Ax$
This is an exact approximation. The second order approximation can perfectly capture a second order polynomial. It should also be noted that the approximation has the exact same formula as the original function showing that it is also equivalent.

c) What are the necessary and sufficient conditions for A to positive definite?
For A to be positive definite, all eigen values of A must be positive

d) What are the necessary and sufficient conditions for A to have full rank?
For a matrix to have full rank, each row is linearly independent. Each column is also independent. This can only happen if the matrix has a non-zero determinant. A would not be singular.

e)If there exists $y \in \mathbb{R}^n \ $ and $y \neq 0 \ $ such that $A^Ty = 0 \ $ then what are the conditions on $ b \ for \ Ax=b \ $ to have a solution for $x$?
Since $A^Ty = 0 \ $ and $ y \neq 0 \ $ , y must be orthogonal to A. Oppositely b must be within the same space as A inorder for $Ax=b$. This means that b is in the space of A and is perpendicular to y

## Problem 3
Stigler Diet Problem
$ \imath = 1,2,...,N \ $is all the different foods at the store
$ \jmath = 1,2,...,M \ $is all nutrients in a food
$ a_{\imath \jmath} \ $ is the quantity of nutrition type $\jmath \ $ in each food $\imath \ $
$ c_\imath \ $ is the unit price of food type $\imath \ $ and $b_\jmath \ $ is the necessary quantity of nutrition type $\jmath \ $ for a month.
x is a column vector of the quantity of each food bought

minimize: $c^Tx$ <br>
subject to: $a^Tx = b$
<nbtab>$ 0 \leq x$

